### PR TITLE
[google-cloud-cpp] fix grpc_cpp_plugin on macOS, shared

### LINKED ISSUE #23266: [package] google-cloud-cpp/2.x: broken on conan v1 with `os=macOS`, `shared=True`
### Description

Running a build at HEAD produces the following errors: https://github.com/conan-io/conan-center-index/pull/23149#issuecomment-2004990402
https://c3i.jfrog.io/c3i/misc/summary.html?json=https://c3i.jfrog.io/c3i/misc/logs/pr/23149/1-macos-m1-clang/google-cloud-cpp/2.12.0//summary.json

The relevant settings are:
- conan v1
- macOS
- shared=True

It fails when trying to run the `grpc_cpp_plugin`. It cannot find the library `libprotoc`. I assume this has to do with the SIP issues clearing the `DYLD_LIBRARY_PATH` on macOS.

This is blocking the upgrade `google-cloud-cpp` to `2.22.0`.

### Package and Environment Details

google-cloud-cpp

### Conan profile

Configuration (profile_host):
[settings]
arch=armv8
build_type=Debug
compiler=apple-clang
compiler.libcxx=libc++
compiler.version=13.0
os=Macos
[options]
google-cloud-cpp:shared=True
[build_requires]
[env]
[conf]
tools.system.package_manager:mode=install
tools.system.package_manager:sudo=True
tools.apple:sdk_path=/Applications/conan/xcode/13.0/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk

Configuration (profile_build):
[settings]
arch=armv8
arch_build=armv8
build_type=Release
compiler=apple-clang
compiler.libcxx=libc++
compiler.version=13
os=Macos
os_build=Macos
[options]
[build_requires]
[env]

### Steps to reproduce

I cannot reproduce this locally, but modifying the `google-cloud-cpp` (e.g. by adding a comment) and opening a PR at HEAD will run the CI which reveals the problem.

### Logs

https://c3i.jfrog.io/c3i/misc/logs/pr/23149/1-macos-m1-clang/google-cloud-cpp/2.12.0//11eaac8f5ef0ed709285f13c35d5470ac3d26791-build.txt

```
dyld[12597]: Library not loaded: @rpath/libprotoc.32.dylib
  Referenced from: <AC7827B1-65E8-319B-AA17-CE18CCA918F8> /Users/jenkins/workspace/prod-v1/bsr/3770/abcfa/.conan/data/grpc/1.54.3/_/_/package/08b363a449a446ccba121b62ee45677c9adb5c1c/bin/grpc_cpp_plugin
  Reasodyld[12621]: dyld[12631]: Library not loaded: @rpath/libprotoc.32.dylib
  Referenced from: <AC7827B1-65E8-319B-AA17-CE18CCA918F8> /Users/jenkins/workspace/prod-v1/bsr/3770/abcfa/.conan/data/grpc/1.54.3/_/_/package/08b363a449a446ccba121b62ee45677c9adbLibrary not loaded: @rpath/libprdyld[12626]: n: tried: '/Users/jenkins/workspdyld[12622]: dyld[12633]: dyld[12638]: Library not loaded: @rpath/libpr5c1c/bin/grpc_cpp_plugin
  ReasoLibrary not loaded: @rpath/libprLibrary not loaded: @rpath/libprotoc.32.dylib
  Referenced from:Library not loaded: @rpath/libprdyld[12630]: otoc.32.dylib
  Referenced from: <AC7827B1-65E8-319B-AA17-CE18CCLibrary not loaded: @rpath/libprotoc.32.dylib
```

### PR DESCRIPTION
**Pull request by dbolduc**
Specify library name and version:  **google-cloud-cpp/2.x**

Fixes #23266.  I am copying (without understanding) how `grpc` uses `protobuf`. More specifically, I am copying the changes from #19972.

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/google-cloud-cpp/2.x/conanfile.py
Comment (coryan) [inline recipes/google-cloud-cpp/2.x/conanfile.py:165]:
```suggestion
        if not self._is_legacy_one_profile:
            self.tool_requires("grpc/<host_version>")
```

File: recipes/google-cloud-cpp/2.x/conanfile.py
Comment (coryan) [inline recipes/google-cloud-cpp/2.x/conanfile.py:176]:
```suggestion
        VirtualBuildEnv(self).generate()
        if self._is_legacy_one_profile:
            VirtualRunEnv(self).generate(scope="build")
```

**Review by coryan — APPROVED** [review]
The Open Source Python indentation is supposed to be 4 spaces.

**Comment by dbolduc** [comment]
> The Open Source Python indentation is supposed to be 4 spaces.

Ack. `google-cloud-cpp` has another pending PR that is blocked on this one (#22993). I will correct the indents in that PR. That should get us through the review process more quickly.

**Comment by RubenRBS** [comment]
@dbolduc do you mean that the linked PR is the canonical one? Or should we review & merge this too?

**Comment by dbolduc** [comment]
> do you mean that the linked PR is the canonical one? Or should we review & merge this too?

@RubenRBS  - This PR needs to go first. This one fixes the `google-cloud-cpp` recipe at HEAD.

In the linked PR we will add the latest version of `google-cloud-cpp`. That PR is not ready for review yet. I will have to rebase it after this PR goes through.

**Comment by dbolduc** [comment]
@RubenRBS ping?

File: recipes/google-cloud-cpp/2.x/conanfile.py
Comment (dbolduc) [reply recipes/google-cloud-cpp/2.x/conanfile.py]: Fixed

File: recipes/google-cloud-cpp/2.x/conanfile.py
Comment (dbolduc) [reply recipes/google-cloud-cpp/2.x/conanfile.py]: Fixed

File: recipes/google-cloud-cpp/2.x/test_package/conanfile.py
Comment (RubenRBS) [inline recipes/google-cloud-cpp/2.x/test_package/conanfile.py:19]: My only concern now is that the test package also needs the dependency, not opposing it, but having a bit more insight into why this is the case would be great!

**Review by RubenRBS — COMMENTED** [review]
Thanks a lot @dbolduc, sorry about the delay :)

Only 1 minor question before merging, most of the changes look good

File: recipes/google-cloud-cpp/2.x/test_package/conanfile.py
Comment (dbolduc) [reply recipes/google-cloud-cpp/2.x/test_package/conanfile.py]: Good catch. Removed.

I was copying gRPC who run protobuf in their test package:

https://github.com/SpaceIm/conan-center-index/blob/9991d8f1f7c29bd7e7ee8ea40bf4cac593fc8147/recipes/grpc/all/test_package/CMakeLists.txt#L21-L30

But I don't think `google-cloud-cpp` needs anything like that.

**Comment by dbolduc** [comment]
> sorry about the delay :)

No worries. Seems like we will have to wait for the CI to come back online anyway.

**Review by RubenRBS — APPROVED** [review]

**Review by toge — APPROVED** [review]

**Comment by uilianries** [comment]
This recipe always used gprc and protobuf, and passed with Conan 1.x before. Why is it failing only now?

**Comment by dbolduc** [comment]
> This recipe always used gprc and protobuf, and passed with Conan 1.x before. Why is it failing only now?

I am 95% sure that the `grpc` update in #19972 broke `google-cloud-cpp`.

**Pull request merged**
